src/clr-angular/forms/password/password-container.spec.ts

```diff
@@ -0,0 +1,184 @@
+    </clr-password-container>`,
+})
+class NoLabelTest {}
+
+export default function(): void {
+  describe('ClrPasswordContainer', () => {
+    let fixture, containerDE, container, containerEl, ifErrorService, layoutService, focusService;
+
+    describe('no label', () => {
+      beforeEach(() => {
+        TestBed.configureTestingModule({
+          imports: [ClrIconModule, ClrCommonFormsModule, FormsModule],
+          declarations: [ClrPasswordContainer, ClrPassword, NoLabelTest],
+          providers: [NgControl, NgControlService, IfErrorService, LayoutService],
+        });
+        fixture = TestBed.createComponent(NoLabelTest);
+
+        containerDE = fixture.debugElement.query(By.directive(ClrPasswordContainer));
+        containerEl = containerDE.nativeElement;
+        layoutService = containerDE.injector.get(LayoutService);
+      });
+
+      it('adds an empty label when instantiated without vertical layout', () => {
+        layoutService.layout = Layouts.HORIZONTAL;
+        fixture.detectChanges();
+        const labels = containerEl.querySelectorAll('label');
+        expect(Array.prototype.filter.call(labels, label => label.textContent === '').length).toBe(1);
+      });
+    });
+
+    describe('full example', () => {
+      beforeEach(() => {
+        TestBed.configureTestingModule({
+          imports: [ClrIconModule, ClrCommonFormsModule, FormsModule],
+          declarations: [ClrPasswordContainer, ClrPassword, SimpleTest],
+          providers: [NgControl, NgControlService, IfErrorService, LayoutService],
+        });
+        fixture = TestBed.createComponent(SimpleTest);
+
+        containerDE = fixture.debugElement.query(By.directive(ClrPasswordContainer));
+        container = containerDE.componentInstance;
+        containerEl = containerDE.nativeElement;
+        ifErrorService = containerDE.injector.get(IfErrorService);
+        layoutService = containerDE.injector.get(LayoutService);
+        focusService = containerDE.injector.get(FocusService);
+        fixture.detectChanges();
+      });
+
+      it('injects the IfErrorService and FocusService and subscribes in AfterContentInit', () => {
+        expect(ifErrorService).toBeTruthy();
+        expect(ifErrorService).toBeTruthy();
+        expect(container.subscriptions.length).toEqual(2);
+      });
+
+      it('projects the label as first child', () => {
+        const label = containerEl.querySelector('label');
+        expect(label).toBeTruthy();
+        expect(label.previousElementSibling).toBeFalsy();
+      });
+
+      it('projects the password inside of the input-wrapper', () => {
+        expect(containerEl.querySelector('.clr-input-wrapper [clrPassword]')).toBeTruthy();
+      });
+
+      it('projects the helper text when the password is valid', () => {
+        expect(containerEl.querySelector('clr-control-helper')).toBeTruthy();
+      });
+
+      it("doesn't display the helper text when invalid", () => {
```

Comment:
Note to self: we could really use some generic specs to reuse for every input component.

Reply:
done